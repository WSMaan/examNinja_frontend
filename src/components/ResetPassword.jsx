import React, { useState } from 'react';
import { Box, TextField, Button, Typography, Alert, Stack } from '@mui/material';
import { Formik, Form } from 'formik';
import * as Yup from 'yup';
import { resetPassword } from '../services/APIservice.jsx'; // API call for reset password

const validationSchema = Yup.object({
  email: Yup.string()
    .email('Enter a valid email address')
    .required('Email is mandatory'),
  password: Yup.string()
    .min(8, 'Password must be at least 8 characters')
    .max(15, 'Password must not exceed 15 characters')
    .matches(/[A-Z]/, 'Password must contain at least one uppercase letter')
    .matches(/[0-9]/, 'Password must contain at least one number')
    .matches(/[!@#$%^&*(),.?":{}|<>]/, 'Password must contain at least one special character')
    .required('Password is required'),
  confirmPassword: Yup.string()
    .oneOf([Yup.ref('password'), null], 'Passwords must match')
    .required('Confirm Password is required'),
});

const ResetPassword = ({ onClose }) => {
  const [successMessage, setSuccessMessage] = useState(null);
  const [errorMessage, setErrorMessage] = useState(null);

  const initialValues = {
    email: '',
    password: '',
    confirmPassword: '',
  };

  const handleSubmit = async (values) => {
    try {
      const response = await resetPassword({
        email: values.email,
        password: values.password, // Only sending email and password
      });
      if (response && response.status === 200) {
        setSuccessMessage('Password updated successfully');
        setErrorMessage(null);
        onClose(); // Close modal after success
      }
    } catch (error) {
      if (error.response?.status === 400) {
        setErrorMessage(error.response.data.error.password || 'Validation failed');
      } else if (error.response?.status === 404) {
        setErrorMessage('User not found with the provided email.');
      } else {
        setErrorMessage('An unexpected error occurred. Please try again.');
      }
      setSuccessMessage(null);
    }
  };

  return (
    <Box>
      <Typography variant="h6" sx={{ mb: 2 }}>
        Reset Password
      </Typography>

      {successMessage && (
        <Stack sx={{ width: '100%', mb: 2 }} spacing={2}>
          <Alert severity="success">{successMessage}</Alert>
        </Stack>
      )}
      {errorMessage && (
        <Stack sx={{ width: '100%', mb: 2 }} spacing={2}>
          <Alert severity="error">{errorMessage}</Alert>
        </Stack>
      )}

      <Formik
        initialValues={initialValues}
        validationSchema={validationSchema}
        onSubmit={handleSubmit}
      >
        {({ errors, touched, handleChange, handleBlur, values }) => (
          <Form>
            <TextField
              fullWidth
              label="Email Address"
              name="email"
              variant="outlined"
              value={values.email}
              onChange={handleChange}
              onBlur={handleBlur}
              error={touched.email && Boolean(errors.email)}
              helperText={touched.email && errors.email}
              sx={{ mb: 2 }}
            />
            <TextField
              fullWidth
              label="New Password"
              name="password"
              type="password"
              variant="outlined"
              value={values.password}
              onChange={handleChange}
              onBlur={handleBlur}
              error={touched.password && Boolean(errors.password)}
              helperText={touched.password && errors.password}
              sx={{ mb: 2 }}
            />
            <TextField
              fullWidth
              label="Confirm Password"
              name="confirmPassword"
              type="password"
              variant="outlined"
              value={values.confirmPassword}
              onChange={handleChange}
              onBlur={handleBlur}
              error={touched.confirmPassword && Boolean(errors.confirmPassword)}
              helperText={touched.confirmPassword && errors.confirmPassword}
              sx={{ mb: 2 }}
            />
            <Button
              type="submit"
              fullWidth
              variant="contained"
              sx={{ mt: 2 }}
            >
              Reset Password
            </Button>
          </Form>
        )}
      </Formik>
    </Box>
  );
};

export default ResetPassword;
