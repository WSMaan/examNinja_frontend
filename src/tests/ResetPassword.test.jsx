import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react'; // Import fireEvent and waitFor
import ResetPassword from '../components/ResetPassword.jsx'; // Adjust the import path as necessary
import axios from 'axios';

// Mock axios
jest.mock('axios');

describe('ResetPassword Component', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<ResetPassword onClose={() => {}} />);
    expect(asFragment()).toMatchSnapshot(); // Create a snapshot
  });

  it('displays error message when email is invalid', async () => {
    const { getByLabelText, getByText } = render(<ResetPassword onClose={() => {}} />);
    
    // Fill in the email field with an invalid email
    const emailInput = getByLabelText(/Email Address/i);
    fireEvent.change(emailInput, { target: { value: 'invalid-email' } });
    fireEvent.blur(emailInput); // Trigger validation

    // Click the Continue button
    const submitButton = getByText(/Continue/i);
    fireEvent.click(submitButton);

    // Check for the error message
    expect(await waitFor(() => getByText(/Invalid email format/i))).toBeInTheDocument();
  });
  
  it('displays generated password on successful submission', async () => {
    // Mock axios to return a successful response
    axios.post.mockResolvedValueOnce({
      status: 200,
      data: { message: 'newPassword123' },
    });

    const { getByLabelText, getByText, findByText } = render(<ResetPassword onClose={() => {}} />);
    
    // Fill in a valid email and submit the form
    const emailInput = getByLabelText(/Email Address/i);
    fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
    fireEvent.blur(emailInput); // Trigger validation
    const submitButton = getByText(/Continue/i);
    fireEvent.click(submitButton);
    
    // Wait for the generated password to be displayed
    expect(await findByText(/newPassword123/i)).toBeInTheDocument();
  });
});
