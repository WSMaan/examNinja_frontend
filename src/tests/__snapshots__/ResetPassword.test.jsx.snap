// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetPassword Component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <h6
      class="MuiTypography-root MuiTypography-h6 css-1r1qvyi-MuiTypography-root"
    >
      Forgot Password
    </h6>
    <form
      action="#"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-itbxsg-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1q964xs-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          Email Address
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-1o0kqxp-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1jk99ih-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            name="email"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1pbc52w"
            >
              <span>
                Email Address
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-1aj6dcf-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Continue
      </button>
    </form>
  </div>
</DocumentFragment>
`;
